### ISSUE TITLE
Make Dataset and Database mutable interfaces

### ISSUE THREAD

**Issue by kalman** [attic-labs/noms#1293]
Dataset has a single mutation method `commit` which returns a new Dataset. This is awkward for a few reasons:
* The original Dataset is no longer valid, calling `commit` on it again will fail. Compare this to other immutable interfaces like NomsList where the original is usable.
* Datastores are documented as the mutable entity in noms. It would be consistent if the interface reflected this - e.g. if you update the "foo" dataset, then it makes sense for the Dataset constructed with "foo" to be updated (though any other Datasets constructed with the same ID will become invalid).
* I think a mutable interface in this case would be easier to work with. You're probably not going to be chaining together multiple calls to `Dataset.commit`, vs you might for `NomsList.append`.
* One might also argue that anything with a ref should be immutable, anything without a ref doesn't need to be.

Datastore has 2 mutation methods `writeValue` and `commit`. The former is a mutable interface. The latter isn't, and it's all a little weird because `commit` calls `writeValue` so it *does* mutate the original object.

A mutable style interface is used in our documentation: https://github.com/attic-labs/noms/blob/master/doc/js-tour.md.

Strawman proposal is to replace:
```
class Dataset {
  // Returns a new Dataset whose head is the new commit.
  async commit(v: valueOrPrimitive, parents?): Promise<Dataset>) { ... }
}
```
with
```
class Dataset {
  // Updates the head commit and returns its RefValue.
  // If commit fails, the Promise is rejected.
  async commit(v: valueOrPrimitive, parents?): Promise<RefValue<Commit>>) { ... }
}
```

And I would like to just remove the `Datastore.commit` method entirely.

**Comment by cmasone-attic** [comment]
We have a new proposal! @rafael-atticlabs / @aboodman should I edit/alter Kalman's comment above to reflect it, add it down here as a new comment, or open a new issue with the new proposal and then close this in favor of that? I kinda prefer the latter as it is non-destructive to the original discussion, but still make it clear what the current state is.

**Comment by rafael-atticlabs** [comment]
the later seems fine to me.

**Comment by cmasone-attic** [comment]
Closed in favor of #2589
